vali_objects/utils/p2p_syncer.py

```diff
@@ -186,110 +160,287 @@ def create_golden(self, trusted_checkpoints: dict):
-                        position_dict = json.loads(new_position.to_json_string())
-                        golden_positions[miner_hotkey]["positions"].append(position_dict)
-                    elif (position_uuid not in seen_positions
-                          and position_counts[position_uuid] != 0):
-                        bt.logging.info(f"Position {position_uuid} only appeared [{position_counts[position_uuid]}/{len(valid_checkpoints)}] times on miner {miner_hotkey}. Skipping")
+                    bt.logging.info(
+                        f"Position {position_uuid} only appeared [{position_counts[position_uuid]}/{len(valid_checkpoints)}] times on miner {miner_hotkey}. Skipping")
+
+        # insert heuristic matched positions back into the golden's positions
+        for position in self.heuristic_resolve_positions(unmatched_positions_matrix):
+            bt.logging.info(f"Position {position['position_uuid']} matched, adding back in")
+            miner_hotkey = position["miner_hotkey"]
+            golden_positions[miner_hotkey]["positions"].append(position)
 
         # Construct golden and convert defaultdict to dict
-        self.golden = {"created_timestamp_ms": time_now,
-                       "hard_snap_cutoff_ms": time_now - 1000 * 60 * 15,
-                       "eliminations": eliminations,
+        self.golden = {"created_timestamp_ms": TimeUtil.now_in_millis(),
+                       "hard_snap_cutoff_ms": TimeUtil.now_in_millis() - 1000 * 60 * 15,
+                       "eliminations": golden_eliminations,
                        "positions": {miner: dict(golden_positions[miner]) for miner in golden_positions}}
-        # temp = {k: len(self.golden['positions'][k]['positions']) for k in self.golden['positions']}  # log the {miner hotkey: number of positions}
-        bt.logging.info("-----------dumping summarized golden data-------------")
-        temp = {}
-        for miner in self.golden['positions']:
+        bt.logging.info(f"Created golden checkpoint: {self.checkpoint_summary(self.golden)}")
+        return True
+
+    def sort_positions(self, miner_hotkey, miner_positions, majority_positions, seen_positions, seen_orders, position_counts, order_counts, order_data):
+        """
+        determine which positions are in the majority, and which ones should be matched up using a heuristic
+        """
+        positions_in_majority = []
+        unmatched_positions = []
+
+        for position in miner_positions["positions"]:
+            position_uuid = position["position_uuid"]
+            # position exists on majority of validators
+            if position_uuid in majority_positions:
+                # create a single combined position, and delete uuid to avoid duplicates
+                new_position = Position(miner_hotkey=miner_hotkey,
+                                        position_uuid=position_uuid,
+                                        open_ms=0,  # position["open_ms"],
```

Comment:
why set open_ms to 0?

Reply:
just a placeholder value, it gets set by the rebuild_position_with_updated_orders() call